ext/datasource/consul/consul.go

```diff
@@ -0,0 +1,161 @@
+package consul
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/alibaba/sentinel-golang/util"
+	"github.com/cenkalti/backoff"
+	"github.com/hashicorp/consul/api"
+)
+
+type consulDataSource struct {
+	datasource.Base
+	*options
+
+	propertyKey string
+	kvQuerier   KVQuerier
+
+	quitCh        chan struct{}
+	isInitialized util.AtomicBool
+	waitIndex     uint64
+}
+
+var (
+	ErrNilConsulClient = errors.New("nil consul client")
+
+	logger = logging.GetDefaultLogger()
+)
+
+// NewConsulDataSource returns new consul datasource instance
+func NewConsulDataSource(propertyKey string, opts ...Option) (datasource.DataSource, error) {
+	var options = evaluateOptions(opts)
+	// if not consul client is specified, initialize from the configuration
+	if options.consulClient == nil {
+		client, err := api.NewClient(options.consulConfig)
+		if err != nil {
+			return nil, err
+		}
+		options.consulClient = client
+	}
+
+	// still no consul client, throw error
+	if options.consulClient == nil {
+		return nil, ErrNilConsulClient
+	}
+	return newConsulDataSource(propertyKey, options), nil
+}
+
+func newConsulDataSource(propertyKey string, options *options) *consulDataSource {
+	ds := &consulDataSource{
+		propertyKey: propertyKey,
+		options:     options,
+		quitCh:      make(chan struct{}),
+		kvQuerier:   options.consulClient.KV(),
+	}
+
+	for _, h := range options.propertyHandlers {
+		ds.AddPropertyHandler(h)
+	}
+	return ds
+}
+
+// ReadSource implement datasource.DataSource interface
+func (c *consulDataSource) ReadSource() ([]byte, error) {
+	pair, meta, err := c.kvQuerier.Get(c.propertyKey, c.queryOptions)
```

Comment:
We might need to check nil for pair and meta.

Reply:
query must be checked but meta is safe.